Updating of driver examples across languages

AndyHeap-NeoTech opened the pull request:
New examples use best practices and most up to date code patterns.

Affected Languages: csharp, go, java, javascript, python

AndyHeap-NeoTech committed:
Updating of driver examples across languages

New examples use best practices and most up to date code patterns.

Affected Languages: csharp, go, java, javascript, python

fbiville commented on code/csharp/Example.cs:
```diff
@@ -1,37 +1,24 @@
-  }
-}
\ No newline at end of file
+await using var session = driver.AsyncSession(o => o.WithDatabase("neo4j"));
```
```suggestion
await using var session = driver.AsyncSession(o => o.WithDatabase("movies"));
```

AndyHeap-NeoTech replied:
Implemented on all 5 languages.

fbiville commented on code/csharp/Example.cs:
```diff
@@ -1,37 +1,24 @@
-              new { movieTitle="The Matrix"});
-      return await r.ToListAsync();
-    });
+var cypherQuery = @"
```
needs to be the movie query by default

AndyHeap-NeoTech replied:
Implemented on all 5 languages.

AndyHeap-NeoTech committed:
Updating the new examples to use the movies database and query

AndyHeap-NeoTech commented on code/csharp/Example.cs:
```diff
@@ -1,37 +1,24 @@
+await using var session = driver.AsyncSession(o => o.WithDatabase("neo4j"));
+var result = await session.ReadTransactionAsync(async tx =>
+{
+    var r = await tx.RunAsync(cypherQuery, new { category = "Dairy Products" });
```
Implemented on all 5 languages.

fbiville commented on code/csharp/Example.cs:
```diff
@@ -1,37 +1,24 @@
-      @"
-      MATCH (m:Movie {title:$movieTitle})<-[:ACTED_IN]-(a:Person) RETURN a.name as actorName
-      ";
+using var driver = GraphDatabase.Driver("neo4j://<HOST>:<BOLTPORT>", AuthTokens.Basic("<USERNAME>", "<PASSWORD>"));
```
sorry, did not catch this earlier, URI and credentials should not change

AndyHeap-NeoTech committed:
Updating the URI and auth credentials across examples.

fbiville commented on code/go/example.go:
```diff
@@ -1,68 +1,49 @@
-	return result, err
-}
-
-func handleClose(closer io.Closer, previousError error) error {
```
note: I purposefully removed this. 
It's the correct way to deal with deferred calls' errors but is super verbose and maybe not much of interest for a homepage example

jexp commented on code/go/example.go:
```diff
@@ -1,68 +1,49 @@
-	for _, result := range results {
-		fmt.Println(result)
+	defer driver.Close()
+	results, err := runQuery(driver, getDairyProducts)
```
we should not have a domain specific function name here

jexp replied:
wouldn't an inline func not be better?

fbiville replied:
`runQuery` is not domain-specific though, `getDairyProducts` is.
Happy to inline `runQuery`

AndyHeap-NeoTech replied:
should be renamed getMovies (or similar) as we are now using the movie Db.

jexp commented on code/java/Example.java:
```diff
@@ -1,35 +1,32 @@
 
-import org.neo4j.driver.*;
-import static org.neo4j.driver.Values.parameters;
+import org.neo4j.driver.AuthTokens;
```
let's keep the * import

jexp commented on code/java/Example.java:
```diff
@@ -1,35 +1,32 @@
+    public static void main(String[] args) {
+        try (var driver = GraphDatabase.driver("neo4j+s://demo.neo4jlabs.com:7687", AuthTokens.basic("mUser", "s3cr3t"));
+             var session = driver.session(SessionConfig.forDatabase("movies"))) {
+            var query = new Query("""
```
can we not use a Query Object?

jexp replied:
tying query string and parameters into one is a really uncommon way as parameters are dynamic and query string is not also it's not idiomatic for java database access

injectives replied:
> can we not use a Query Object?

Could you clarify what you mean please? It seems to use the `Query` object.

robsdedude replied:
I think he wants to ask if it's possible to avoid using a Query object because

> tying query string and parameters into one is a really uncommon way as parameters are dynamic and query string is not also it's not idiomatic for java database access

jexp commented on code/javascript/example.js:
```diff
@@ -1,27 +1,41 @@
 
-const params = {"movieTitle": "The Matrix"};
+  try {
+    const products = await getDairyProducts(driver)
```
no domain specific methods that don't work with the templating
why not just an inline closure instead of the full ceremony?

jexp replied:
in this simple example we can close driver and session together

jexp commented on code/javascript/example.js:
```diff
@@ -1,27 +1,41 @@
-const params = {"movieTitle": "The Matrix"};
+  try {
+    const products = await getDairyProducts(driver)
+    products.forEach(product => console.log(product))
```
products.forEach(console.log)

jexp commented on code/javascript/example.js:
```diff
@@ -1,27 +1,41 @@
+  }
+}
+
+main()
```
I wasn't aware that we need a main in JS?
can we avoid this?

bigmontz replied:
This issue is directed linked with: https://github.com/neo4j-graph-examples/template/pull/31/files#r949406759

We don't need `main` in the JS, but we could not `await` in the top level until `Node 14.8.0`. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#top_level_await

There is two common techniques for solving this:
1. Create an `async function main()` (or other arbitrary name), implement everything using `async/await` inside `main` and then call the function (and tread the promise completion). 
2. Mostly the same, but with an anonymous function. Like:
```javascript
(async function () { 
   // the async/ await stuff
})()
  .then(() => console.log('finished'))
  .catch(err => console.error(err)) // or do a proper error treatment.
```

We could also revert for using only promises, which doesn't look great. Or we could write `top-level` await considering the user will be running newer Node versions (which I prefer)

AndyHeap-NeoTech replied:
We'll leave it as it is then.

jexp commented on code/javascript/example.js:
```diff
@@ -1,27 +1,41 @@
+}
+
+main()
+  .then(() => console.log("Finished"))
```
here we're using the promise based api but above not? we should be consistent

jexp commented on code/python/example.py:
```diff
@@ -1,21 +1,24 @@
-driver = GraphDatabase.driver(
-  "neo4j+s://demo.neo4jlabs.com:7687",
-  auth=basic_auth("mUser", "s3cr3t"))
+def products_of_category(transaction, movie_title_):
```
same again, inline lambda, no domain specific method(name) less ceremony

jexp commented on code/python/example.py:
```diff
@@ -1,21 +1,24 @@
-  auth=basic_auth("mUser", "s3cr3t"))
+def products_of_category(transaction, movie_title_):
+    cypher_query = (
+        "MATCH (m:Movie {title:$movieTitle})<-[:ACTED_IN]-(a:Person)" 
```
no double quotes but triple quote strings please otherwise this breaks the templating and is also not idiomatic

jexp commented on code/python/example.py:
```diff
@@ -1,21 +1,24 @@
+    cypher_query = (
+        "MATCH (m:Movie {title:$movieTitle})<-[:ACTED_IN]-(a:Person)" 
+        "RETURN a.name as actorName")
+    result = transaction.run(cypher_query, movie_title=movie_title_)
```
the parameter names need to be consistent across all examples
won't work like this, needs to be `movieTitle`

jexp commented on code/python/example.py:
```diff
@@ -1,21 +1,24 @@
-'''
+uri = "neo4j+s://demo.neo4jlabs.com:7687"
+auth = ("mUser", "s3cr3t")
+driver = GraphDatabase.driver(uri, auth=auth)
```
single line with inline values, no need for the extra vars

fbiville replied:
I guess it has to do with line width?

AndyHeap-NeoTech replied:
It is to do with line length.

robsdedude replied:
79 chars per line according to PEP8. I don't want us to look like we don't know our Python foo.
https://peps.python.org/pep-0008/#maximum-line-length

jexp commented on code/python/example.py:
```diff
@@ -1,21 +1,24 @@
-    print(record['actorName'])
+    categories = session.read_transaction(movies,
+                                          "The Matrix")
+    for category in categories:
```
no domain specific variable names

jexp commented on code/python/example.py:
```diff
@@ -1,21 +1,24 @@
+        print(category)
 
-driver.close()
+driver.close()
```
can driver not be auto-closes with a `with`?

robsdedude replied:
Just saw, that it is actually documented, just not very prominently. So I'm happy to put it in.

jexp commented on code/python/example.py:
```diff
@@ -1,21 +1,24 @@
-                      movieTitle="The Matrix").data())
-  for record in results:
-    print(record['actorName'])
+    categories = session.read_transaction(movies,
```
there is no movies variable

AndyHeap-NeoTech committed:
Fixing errors

Swapping to the movies query and DB I missed some naming changes.
Changed Java example to use single import statement.

AndyHeap-NeoTech committed:
Changed to triple quotes for multiline string.

bigmontz commented on code/javascript/example.js:
```diff
@@ -1,27 +1,41 @@
-const params = {"movieTitle": "The Matrix"};
+  try {
+    const products = await getMovies(driver)
+    products.forEach(product => console.log(product))
```
```suggestion
    products.forEach(console.log)
```

AndyHeap-NeoTech committed:
Minor change to logging in JS example.

injectives commented on code/java/Example.java:
```diff
@@ -1,35 +1,29 @@
+
+            var result = session.readTransaction(tx -> tx.run(query).stream()
+                    .map(record -> record.get("actorName").asString())
+                    .collect(Collectors.toList()));
```
```suggestion
                    .toList())
```

injectives replied:
```suggestion
                    .toList());
```

injectives commented on code/java/Example.java:
```diff
@@ -1,35 +1,29 @@
-      String cypherQuery =
-        "MATCH (m:Movie {title:$movieTitle})<-[:ACTED_IN]-(a:Person) RETURN a.name as actorName";
+import java.util.Map;
+import java.util.stream.Collectors;
```
```suggestion
```

injectives commented on code/java/Example.java:
```diff
@@ -1,35 +1,29 @@
 // download jars into current directory
 // java -cp "*" Example.java
 
 import org.neo4j.driver.*;
```
```suggestion
import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Record;
import org.neo4j.driver.Session;
import org.neo4j.driver.SessionConfig;
```

fbiville commented:
Python and Go have since been updated to use the "execute query" driver-level API.
This PR is now outdated and will be closed.
